File: not solved/Problem 32.py


class Number:
    def __init__(self, product:str, multiplicant:str, multiplier:str):
        self.product = product
        self.multiplicant = multiplicant
        self.multiplier = multiplier

def is_pandigital(num:Number) -> bool:
    number = str(num.product) + str(num.multiplicant) + str(num.multiplier)
    number_occoured = [0, 0, 0, 0, 0, 0, 0, 0, 0]    # contains elements at positions from 0-8(1-9) each position saves, how often the related number occoured in the number

    for digit in number:
        number_occoured[int(digit)-1] += 1
    
    for num in number_occoured:
        if num != 1:    # every digit(from 1-9) has to be in the number exactly 1 time
            return False
    return True
    
# 98765432
upper_limit = 987654
pandigital_numbers = []

for check_product in range(1, upper_limit+1):
    for multiplicant in range(1, check_product+1):
        if check_product % multiplicant == 0:   # Check if we can dived the number without a "rest"
            num = Number(check_product, multiplicant, int(check_product/multiplicant))   # if we can divide the number - the multiplier is simply:
            if is_pandigital(num):
                #print(check_product, ' ', num.multiplier, ' ', num.multiplicant)
                pandigital_numbers.append(check_product)

pandigital_numbers = list(set(pandigital_numbers)) # make every element unique

for number in pandigital_numbers:
    print(number)
    pass

print()
print("Num of pandigital numbers: " + str(len(pandigital_numbers)))

sum = 0
for num in pandigital_numbers:
    sum += num
print("Sum: ", sum)

File: not solved/Problem 32.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

bool is_pandigital(std::string number) {
    unsigned short number_occoured[10] = {0}; // 0-9 - 10
    for(char digit : number) {
        number_occoured[digit - '0'] += 1;
    }
    for(unsigned short i = 1; i < 9; ++i) { // We dont count the 0
        if (number_occoured[i] != 1)
            return false;
    }
    return true;
}

int main() {
    // 98'765'432
    std::vector<unsigned long> pandigital_products;
    for(unsigned long check_product = 1; check_product <= 98'765'432/2; ++check_product) {
        for (unsigned long multiplicand = 1; multiplicand <= check_product/2+1; ++multiplicand) {
            if(check_product % multiplicand == 0) { // if they are divedable
                unsigned long multiplier = check_product / multiplicand;
                std::string number = std::to_string(check_product) + std::to_string(multiplicand) + std::to_string(multiplier);
                if(is_pandigital(number)) {
                    if(std::find(pandigital_products.begin(), pandigital_products.end(), check_product) == pandigital_products.end()) {
                        pandigital_products.push_back(check_product);
                    }
                }
            }
        }
    }

    std::cout << "Num of pandigital Numbers: " << pandigital_products.size() << '\n';
    /*
    std::cout << "Pandigital Numbers:\n";
    for(unsigned int num : pandigital_products) {
        std::cout << num << '\n';
    }
    */
    std::cout << "\n\n";
    
   std::size_t sum = 0;
   for(unsigned long num : pandigital_products) {
       sum += num;
   }
   std::cout << "Sum of pandigital Numbers: " << sum << '\n';


    std::cout << "Fertig\n";

    return 0;
}